### PR TITLE
Pre-check for jobs and release/push copr triggers

### PR DESCRIPTION
**Pull request by lachmanfrantisek**
- Allow to pre-check the jobs to fail early.
- Use the right configured job when running copr builds.
- Add copr-build for release and push. (related issue: #239)
  - I would like to wait with docs and test it properly on stage.
- Refactor the handlers/events.

### CODE REVIEW

File: packit_service/service/events.py [hunk 1]
```diff
@@ -465,7 +499,7 @@ def __init__(
             self.base_repo_namespace = build.get("repo_namespace")
             https_url = build["https_url"]
 
-        super().__init__(trigger=JobTriggerType.commit, project_url=https_url)
+        super().__init__(trigger=JobTriggerType.pull_request, project_url=https_url)
```
Comment (lachmanfrantisek) [inline packit_service/service/events.py]: This is the problem I want to discuss on the arch meeting:
- We need to differentiate between jobs and the `part of the config which contains the metadata about the job`.
- We need to support mapping from jobs to the `part of the config which contains the metadata about the job`.

File: packit_service/worker/build/build_helper.py [hunk 2]
```diff
@@ -85,20 +105,6 @@ def api(self) -> PackitAPI:
...
-            return self.event.base_ref
-
-        if isinstance(self.event, CoprBuildEvent):
-            if self.event.ref:
-                # FIXME: return commit sha always once on PG
-                return self.event.ref
-            return self.event.commit_sha
-
-        if isinstance(self.event, PullRequestCommentEvent):
-            return self.event.commit_sha
```
Comment (lachmanfrantisek) [inline packit_service/worker/build/build_helper.py]: I've tried to remove this kind of code.

File: packit_service/service/events.py [hunk 3]
```diff
@@ -518,9 +552,10 @@ def pre_check(self):
 
     def get_dict(self, default_dict: Optional[Dict] = None) -> dict:
         d = self.__dict__
-        d.pop("build_pg")
+        build_pg = d.pop("build_pg")
         result = super().get_dict(d)
         result["topic"] = result["topic"].value
+        self.build_pg = build_pg
```
Comment (lachmanfrantisek) [inline packit_service/service/events.py]: Honestly, I am not 100% sure what I am doing here -- why we need to edit `__dict__` here just for the get method?
(I understand, that the `build_pg` is not serializable.)

Comment (TomasTomecek) [reply packit_service/service/events.py]: this is all just for sake of serialization and the way this is implemented atm

File: packit_service/worker/handlers/abstract.py [hunk 4]
```diff
@@ -112,6 +112,14 @@ def _clean_workplace(self):
             else:
                 shutil.rmtree(item)
 
+    def pre_check(self) -> bool:
+        """
+        Validate the input.
+
+        :return: False if we can skip the run.
```
Comment (TomasTomecek) [inline packit_service/worker/handlers/abstract.py]: pre_cache -> False mean that run() will not be executed?

Comment (lachmanfrantisek) [reply packit_service/worker/handlers/abstract.py]: Currently only for the configured jobs (in the `process_jobs` method), not the hardcoded ones in the `process_message`. Shall I use that also there?

File: packit_service/service/events.py (hunk 3 above)
Comment (lachmanfrantisek) [reply packit_service/service/events.py]: Is this change safe? (I was having weird errors in tests because of the missing attribute `build_pg`...)

**Comment by lachmanfrantisek** [comment]
I will merge it to play with that on stage tomorrow. (So I have time to fix potential problems until Friday..;)

Feel free to continue with the review.

File: packit_service/worker/handlers/github_handlers.py [hunk 5]
```diff
@@ -235,87 +235,168 @@ def run(self) -> HandlerResults:
...
+    ):
+        super().__init__(config=config, job=job, event=event)
+
+    def run(self) -> HandlerResults:
+        if isinstance(self.event, GithubPullRequestHandler):
+            collaborators = self.project.who_can_merge_pr()
+            if self.event.github_login not in collaborators | self.config.admins:
+                msg = (
+                    "Only users with write or admin permissions to the repository "
+                    "can trigger Packit-as-a-Service"
```
Comment (jpopelka) [inline packit_service/worker/handlers/github_handlers.py]: I've seen lots of those in Hunor's PR. Can we refactor them into a constant? for example `ADMINS_ONLY`

File: packit_service/worker/handlers/abstract.py (hunk 4 above)
Comment (TomasTomecek) [reply packit_service/worker/handlers/abstract.py]: I'd love to see this text in the docstring - the present docstring is really sparse.

File: packit_service/service/events.py (hunk 3 above)
Comment (TomasTomecek) [reply packit_service/service/events.py]: should be fine

File: packit_service/worker/handlers/github_handlers.py (hunk 5 above)
Comment (lachmanfrantisek) [reply packit_service/worker/handlers/github_handlers.py]: Fixed in #451

File: packit_service/worker/handlers/abstract.py (hunk 4 above)
Comment (lachmanfrantisek) [reply packit_service/worker/handlers/abstract.py]: I've updated the doc string in #451

Comment (lachmanfrantisek) [reply packit_service/worker/handlers/abstract.py]: I forgot to push that commit, will be in the next one.

**Pull request merged**